File: prani-backend/src/main/java/com/prani/service/AuthService.java
package com.prani.service;

import com.prani.entity.EmailOtp;
import com.prani.entity.User;
import com.prani.entity.UserSession;
import com.prani.exception.ResourceNotFoundException;
import com.prani.exception.UnauthorizedException;
import com.prani.repository.EmailOtpRepository;
import com.prani.repository.UserRepository;
import com.prani.repository.UserSessionRepository;
import com.prani.security.JwtService;
import dev.samstevens.totp.code.CodeGenerator;
import dev.samstevens.totp.code.DefaultCodeGenerator;
import dev.samstevens.totp.code.DefaultCodeVerifier;
import dev.samstevens.totp.code.CodeVerifier;
import dev.samstevens.totp.secret.DefaultSecretGenerator;
import dev.samstevens.totp.time.SystemTimeProvider;
import dev.samstevens.totp.time.TimeProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final UserSessionRepository userSessionRepository;
    private final EmailOtpRepository emailOtpRepository;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final EmailService emailService;
    private final AuditService auditService;
    private final SubscriptionService subscriptionService;

    private final TimeProvider timeProvider = new SystemTimeProvider();
    private final CodeGenerator codeGenerator = new DefaultCodeGenerator();
    private final CodeVerifier codeVerifier = new DefaultCodeVerifier(codeGenerator, timeProvider);

    @Transactional
    public Map<String, Object> register(String email, String password, String name, String role) {
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already registered");
        }

        String userId = UUID.randomUUID().toString();
        User user = User.builder()
            .userId(userId)
            .email(email.toLowerCase().trim())
            .name(name)
            .role(role != null ? role : "client")
            .passwordHash(passwordEncoder.encode(password))
            .mfaEnabled(false)
            .isActive(true)
            .skills(List.of())
            .certifications(List.of())
            .build();

        userRepository.save(user);

        // Create free subscription for new user
        subscriptionService.createFreeSubscription(userId);

        String token = jwtService.createToken(userId, user.getRole(), user.getEmail());
        return Map.of("token", token, "user", toUserMap(user));
    }

    public Map<String, Object> login(String email, String password, String ipAddress) {
        User user = userRepository.findByEmail(email.toLowerCase().trim())
            .orElseThrow(() -> new UnauthorizedException("Invalid credentials"));

        if (!user.getIsActive()) {
            throw new UnauthorizedException("Account is deactivated");
        }
        if (user.getPasswordHash() == null ||
            !passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid credentials");
        }

        if (Boolean.TRUE.equals(user.getMfaEnabled())) {
            return Map.of("mfa_required", true, "user_id", user.getUserId());
        }

        String token = jwtService.createToken(user.getUserId(), user.getRole(), user.getEmail());
        auditService.log(null, "login", "user", user.getUserId(),
            Map.of("email", email), ipAddress);
        return Map.of("token", token, "user", toUserMap(user));
    }

    @Transactional
    public Map<String, Object> verifyMfa(String userId, String code, String method, String ipAddress) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        boolean valid = false;

        if ("totp".equals(method)) {
            valid = codeVerifier.isValidCode(user.getMfaSecret(), code);
        } else if ("email".equals(method)) {
            EmailOtp otp = emailOtpRepository.findTopByUserIdOrderByCreatedAtDesc(userId)
                .orElseThrow(() -> new UnauthorizedException("No OTP found"));
            if (otp.getExpiresAt().isAfter(OffsetDateTime.now()) && otp.getCode().equals(code)) {
                valid = true;
                emailOtpRepository.delete(otp);
            }
        }

        if (!valid) {
            throw new UnauthorizedException("Invalid or expired code");
        }

        String token = jwtService.createToken(user.getUserId(), user.getRole(), user.getEmail());
        auditService.log(null, "mfa_login", "user", userId, Map.of("method", method), ipAddress);
        return Map.of("token", token, "user", toUserMap(user));
    }

    @Transactional
    public void sendEmailOtp(String userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        String code = String.format("%06d", new Random().nextInt(999999));
        EmailOtp otp = EmailOtp.builder()
            .otpId(UUID.randomUUID().toString())
            .userId(userId)
            .code(code)
            .expiresAt(OffsetDateTime.now().plusMinutes(10))
            .build();
        emailOtpRepository.save(otp);
        emailService.sendOtpEmail(user.getEmail(), code);
    }

    public Map<String, String> setupTotp(String userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        String secret = new DefaultSecretGenerator().generate();
        String otpAuthUri = "otpauth://totp/Prani:" + user.getEmail()
            + "?secret=" + secret + "&issuer=Prani";

        // Store secret temporarily — confirmed on verify-totp-setup
        user.setMfaSecret(secret);
        userRepository.save(user);

        return Map.of("secret", secret, "otp_uri", otpAuthUri);
    }

    @Transactional
    public void verifyTotpSetup(String userId, String code) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        if (!codeVerifier.isValidCode(user.getMfaSecret(), code)) {
            throw new UnauthorizedException("Invalid TOTP code");
        }

        user.setMfaEnabled(true);
        userRepository.save(user);
        auditService.log(null, "mfa_setup", "user", userId, Map.of("method", "totp"), null);
    }

    @Transactional
    public void changePassword(String userId, String currentPassword, String newPassword) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        if (!passwordEncoder.matches(currentPassword, user.getPasswordHash())) {
            throw new UnauthorizedException("Current password is incorrect");
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        userRepository.save(user);
        auditService.log(null, "password_change", "user", userId, Map.of(), null);
    }

    // Google OAuth — lookup session token, return JWT (mirrors Python POST /auth/google)
    @Transactional
    public Map<String, Object> loginWithSession(String sessionId, String ipAddress) {
        UserSession session = userSessionRepository.findBySessionToken(sessionId)
            .orElseThrow(() -> new UnauthorizedException("Invalid or expired session"));

        User user = userRepository.findById(session.getUserId())
            .orElseThrow(() -> new UnauthorizedException("User not found"));

        if (!user.getIsActive()) {
            throw new UnauthorizedException("Account is deactivated");
        }

        String token = jwtService.createToken(user.getUserId(), user.getRole(), user.getEmail());
        auditService.log(null, "google_login", "user", user.getUserId(), Map.of(), ipAddress);
        return Map.of("token", token, "user", toUserMap(user));
    }

    public Map<String, Object> toUserMap(User user) {
        return Map.of(
            "user_id", user.getUserId(),
            "email", user.getEmail(),
            "name", user.getName(),
            "role", user.getRole(),
            "mfa_enabled", Boolean.TRUE.equals(user.getMfaEnabled()),
            "is_active", Boolean.TRUE.equals(user.getIsActive()),
            "skills", user.getSkills() != null ? user.getSkills() : List.of(),
            "certifications", user.getCertifications() != null ? user.getCertifications() : List.of(),
            "created_at", user.getCreatedAt() != null ? user.getCreatedAt().toString() : null
        );
    }
}

File: node-backend/src/services/auth.service.js
'use strict';

const bcrypt = require('bcryptjs');
const { randomUUID } = require('crypto');
const prisma = require('../config/prisma');
const { createToken } = require('../utils/jwt');
const { generateSecret, generateQrUri, verify: verifyTotp, generateEmailOtp } = require('../utils/totp');
const { AppError } = require('../middleware/errorHandler');
const emailService = require('./email.service');
const auditService = require('./audit.service');
const subscriptionService = require('./subscription.service');
const config = require('../config/env');

const BCRYPT_ROUNDS = 10;
const EMAIL_OTP_TTL_MINUTES = 10;

// ─── Helpers ─────────────────────────────────────────────────────────────────

/**
 * Strip sensitive fields before returning a user to the caller.
 */
function safeUser(user) {
  const { passwordHash, mfaSecret, ...rest } = user;
  return rest;
}

/**
 * Build a standard auth response object.
 */
function buildAuthResponse(user, token) {
  return { token, user: safeUser(user) };
}

// ─── Service methods ──────────────────────────────────────────────────────────

/**
 * Register a new user account.
 *
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.password
 * @param {string} params.name
 * @param {string} [params.role]
 * @param {string} [params.tenantId]
 * @returns {Promise<{ token: string, user: object }>}
 */
async function register({ email, password, name, role = 'client', tenantId = null }) {
  const existing = await prisma.user.findUnique({ where: { email: email.toLowerCase().trim() } });
  if (existing) {
    throw new AppError('An account with this email address already exists', 409, 'EMAIL_TAKEN');
  }

  const passwordHash = await bcrypt.hash(password, BCRYPT_ROUNDS);

  const user = await prisma.user.create({
    data: {
      email: email.toLowerCase().trim(),
      passwordHash,
      name,
      role,
      tenantId,
      isActive: true,
    },
  });

  try {
    await subscriptionService.createAutoTrial(user.userId, user.tenantId || null);
  } catch (trialErr) {
    // Trial creation may fail if DB constraints haven't been migrated yet.
    // Fall back silently to free plan — user can always upgrade later.
    console.warn('[register] createAutoTrial failed, falling back to free:', trialErr.message);
    await subscriptionService.createFreeSubscription(user.userId, user.tenantId || null).catch(() => {});
  }

  const token = createToken(user.userId, user.role, user.email, user.tenantId);

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.REGISTER,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
  });

  await emailService.sendWelcome(user.email, user.name).catch(() => {});

  return buildAuthResponse(user, token);
}

/**
 * Authenticate a user with email + password.
 * If MFA is enabled, returns a challenge response instead of a token.
 *
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.password
 * @param {string} [params.ipAddress]
 * @returns {Promise<{ token: string, user: object } | { mfa_required: true, userId: string, method: string }>}
 */
async function login({ email, password, ipAddress }) {
  const user = await prisma.user.findUnique({ where: { email: email.toLowerCase().trim() } });
  if (!user) {
    throw new AppError('Invalid email or password', 401, 'INVALID_CREDENTIALS');
  }

  if (!user.isActive) {
    throw new AppError('Your account has been deactivated. Please contact support.', 403, 'ACCOUNT_INACTIVE');
  }

  const passwordMatch = await bcrypt.compare(password, user.passwordHash);
  if (!passwordMatch) {
    throw new AppError('Invalid email or password', 401, 'INVALID_CREDENTIALS');
  }

  if (user.mfaEnabled) {
    return {
      mfa_required: true,
      userId: user.userId,
      method: 'totp',
    };
  }

  const token = createToken(user.userId, user.role, user.email, user.tenantId);

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.LOGIN,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    ipAddress,
  });

  return buildAuthResponse(user, token);
}

/**
 * Verify an MFA challenge (TOTP or email OTP) and issue a JWT on success.
 *
 * @param {object} params
 * @param {string} params.userId
 * @param {string} params.code
 * @param {string} params.method  - 'totp' | 'email'
 * @param {string} [params.ipAddress]
 * @returns {Promise<{ token: string, user: object }>}
 */
async function verifyMfa({ userId, code, method, ipAddress }) {
  const user = await prisma.user.findUnique({ where: { userId } });
  if (!user || !user.isActive) {
    throw new AppError('User not found or inactive', 404, 'USER_NOT_FOUND');
  }

  let verified = false;

  if (method === 'totp') {
    if (!user.mfaSecret) {
      throw new AppError('TOTP is not configured for this account', 400, 'TOTP_NOT_CONFIGURED');
    }
    verified = verifyTotp(user.mfaSecret, code);
  } else if (method === 'email') {
    const otpRecord = await prisma.emailOtp.findFirst({
      where: {
        userId,
        expiresAt: { gt: new Date() },
      },
      orderBy: { createdAt: 'desc' },
    });

    if (!otpRecord) {
      throw new AppError('OTP not found or has expired. Please request a new code.', 400, 'OTP_EXPIRED');
    }

    verified = otpRecord.code === code;

    if (verified) {
      await prisma.emailOtp.delete({ where: { id: otpRecord.id } });
    }
  } else {
    throw new AppError(`Unsupported MFA method: ${method}`, 400, 'INVALID_MFA_METHOD');
  }

  if (!verified) {
    throw new AppError('Invalid verification code', 401, 'INVALID_MFA_CODE');
  }

  const token = createToken(user.userId, user.role, user.email, user.tenantId);

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.MFA_VERIFY,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    details: { method },
    ipAddress,
  });

  return buildAuthResponse(user, token);
}

/**
 * Generate and send an email OTP to the user.
 *
 * @param {string} userId
 * @returns {Promise<void>}
 */
async function sendEmailOtp(userId) {
  const user = await prisma.user.findUnique({ where: { userId } });
  if (!user || !user.isActive) {
    throw new AppError('User not found', 404, 'USER_NOT_FOUND');
  }

  const code = generateEmailOtp();
  const expiresAt = new Date(Date.now() + EMAIL_OTP_TTL_MINUTES * 60 * 1000);

  // Remove any existing OTPs for this user before creating a new one
  await prisma.emailOtp.deleteMany({ where: { userId } });

  await prisma.emailOtp.create({
    data: {
      userId,
      code,
      expiresAt,
    },
  });

  await emailService.sendEmailOtp(user.email, code);
}

/**
 * Initiate TOTP setup: generate a new secret and return the QR URI.
 * The secret is saved to the user but MFA is NOT yet enabled.
 *
 * @param {string} userId
 * @returns {Promise<{ qrUri: string, secret: string }>}
 */
async function setupTotp(userId) {
  const user = await prisma.user.findUnique({ where: { userId } });
  if (!user) {
    throw new AppError('User not found', 404, 'USER_NOT_FOUND');
  }

  const secret = generateSecret();
  const qrUri = generateQrUri(user.email, secret);

  await prisma.user.update({
    where: { userId },
    data: { mfaSecret: secret },
  });

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.MFA_SETUP,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    details: { step: 'initiated' },
  });

  return { qrUri, secret };
}

/**
 * Verify the first TOTP code to confirm setup and enable MFA on the account.
 *
 * @param {object} params
 * @param {string} params.userId
 * @param {string} params.code
 * @returns {Promise<{ success: true }>}
 */
async function verifyTotpSetup({ userId, code }) {
  const user = await prisma.user.findUnique({ where: { userId } });
  if (!user || !user.mfaSecret) {
    throw new AppError('TOTP setup has not been initiated. Call /totp-setup first.', 400, 'TOTP_NOT_INITIATED');
  }

  const valid = verifyTotp(user.mfaSecret, code);
  if (!valid) {
    throw new AppError('Invalid TOTP code. Please try again.', 401, 'INVALID_TOTP_CODE');
  }

  await prisma.user.update({
    where: { userId },
    data: { mfaEnabled: true },
  });

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.MFA_SETUP,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    details: { step: 'completed' },
  });

  return { success: true };
}

/**
 * Change the authenticated user's password.
 *
 * @param {object} params
 * @param {string} params.userId
 * @param {string} params.currentPassword
 * @param {string} params.newPassword
 * @returns {Promise<{ success: true }>}
 */
async function changePassword({ userId, currentPassword, newPassword }) {
  const user = await prisma.user.findUnique({ where: { userId } });
  if (!user) {
    throw new AppError('User not found', 404, 'USER_NOT_FOUND');
  }

  const match = await bcrypt.compare(currentPassword, user.passwordHash);
  if (!match) {
    throw new AppError('Current password is incorrect', 401, 'WRONG_PASSWORD');
  }

  const newHash = await bcrypt.hash(newPassword, BCRYPT_ROUNDS);

  await prisma.user.update({
    where: { userId },
    data: { passwordHash: newHash },
  });

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.PASSWORD_CHANGE,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
  });

  await emailService.sendPasswordChanged(user.email, user.name).catch(() => {});

  return { success: true };
}

/**
 * Authenticate via a persisted session token (cookie-based session).
 *
 * @param {object} params
 * @param {string} params.sessionToken
 * @param {string} [params.ipAddress]
 * @returns {Promise<{ token: string, user: object }>}
 */
async function loginWithSession({ sessionToken, ipAddress }) {
  const session = await prisma.userSession.findUnique({
    where: { sessionToken },
    include: {
      user: true,
    },
  });

  if (!session) {
    throw new AppError('Session not found', 401, 'INVALID_SESSION');
  }

  if (session.expiresAt < new Date()) {
    throw new AppError('Session has expired', 401, 'SESSION_EXPIRED');
  }

  const user = session.user;

  if (!user || !user.isActive) {
    throw new AppError('Account not found or inactive', 401, 'ACCOUNT_INACTIVE');
  }

  const token = createToken(user.userId, user.role, user.email, user.tenantId);

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.LOGIN,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    details: { method: 'session' },
    ipAddress,
  });

  return buildAuthResponse(user, token);
}

/**
 * Accept a tenant invitation, create the user account, and return a JWT.
 *
 * @param {object} params
 * @param {string} params.token    - Invitation token from the email link
 * @param {string} params.password - Password chosen by the invitee
 * @param {string} params.name     - Display name chosen by the invitee
 * @returns {Promise<{ token: string, user: object }>}
 */
async function acceptInvitation({ token, password, name }) {
  const invitation = await prisma.tenantInvitation.findUnique({ where: { token } });

  if (!invitation) {
    throw new AppError('Invitation not found', 404, 'INVITATION_NOT_FOUND');
  }

  if (invitation.expiresAt < new Date()) {
    throw new AppError('This invitation has expired. Please ask for a new one.', 400, 'INVITATION_EXPIRED');
  }

  if (invitation.acceptedAt) {
    throw new AppError('This invitation has already been accepted', 400, 'INVITATION_ALREADY_ACCEPTED');
  }

  const existing = await prisma.user.findUnique({ where: { email: invitation.email.toLowerCase() } });
  if (existing) {
    throw new AppError('An account with this email address already exists', 409, 'EMAIL_TAKEN');
  }

  const passwordHash = await bcrypt.hash(password, BCRYPT_ROUNDS);

  const user = await prisma.user.create({
    data: {
      email: invitation.email.toLowerCase(),
      passwordHash,
      name,
      role: invitation.role,
      tenantId: invitation.tenantId,
      isActive: true,
    },
  });

  await subscriptionService.createFreeSubscription(user.userId);

  await prisma.tenantInvitation.update({
    where: { id: invitation.id },
    data: { acceptedAt: new Date() },
  });

  const jwtToken = createToken(user.userId, user.role, user.email, user.tenantId);

  await auditService.log({
    userId: user.userId,
    userEmail: user.email,
    action: auditService.AuditAction.REGISTER,
    resource: auditService.AuditResource.AUTH,
    resourceId: user.userId,
    details: { via: 'invitation', tenantId: invitation.tenantId },
  });

  await emailService.sendWelcome(user.email, user.name).catch(() => {});

  return buildAuthResponse(user, jwtToken);
}

/**
 * Retrieve invitation details by token for pre-filling the acceptance form.
 *
 * @param {string} token
 * @returns {Promise<object>} The invitation record
 */
async function getInvitation(token) {
  const invitation = await prisma.tenantInvitation.findUnique({ where: { token } });

  if (!invitation) {
    throw new AppError('Invitation not found', 404, 'INVITATION_NOT_FOUND');
  }

  if (invitation.expiresAt < new Date()) {
    throw new AppError('This invitation has expired', 400, 'INVITATION_EXPIRED');
  }

  if (invitation.acceptedAt) {
    throw new AppError('This invitation has already been accepted', 400, 'INVITATION_ALREADY_ACCEPTED');
  }

  return invitation;
}

/**
 * Initiate a self-service password reset. Always returns without error to
 * prevent email enumeration — the reset email is sent only when the account exists.
 *
 * @param {string} email
 */
async function forgotPassword(email) {
  const normalised = email.toLowerCase().trim();
  const user = await prisma.user.findUnique({
    where: { email: normalised },
    select: { userId: true, name: true, email: true, isActive: true },
  });

  if (!user || !user.isActive) return; // silent — don't reveal whether email exists

  const token = randomUUID();
  const expiresAt = new Date(Date.now() + 60 * 60 * 1000); // 1 hour

  // Invalidate any existing tokens for this user
  await prisma.$executeRawUnsafe(
    `DELETE FROM password_reset_tokens WHERE user_id = $1`,
    user.userId
  );

  await prisma.$executeRawUnsafe(
    `INSERT INTO password_reset_tokens (user_id, token, expires_at)
     VALUES ($1, $2, $3::timestamptz)`,
    user.userId, token, expiresAt.toISOString()
  );

  const resetLink = `${config.frontendUrl}/reset-password?token=${token}`;
  await emailService.sendPasswordResetRequest(user.email, user.name, resetLink).catch((err) => {
    console.error('[auth.service] Failed to send password reset email:', err.message);
  });
}

/**
 * Complete a password reset using the token sent by email.
 *
 * @param {string} token       - UUID token from the reset link
 * @param {string} newPassword - New plaintext password (min 8 chars)
 */
async function resetPassword(token, newPassword) {
  const rows = await prisma.$queryRawUnsafe(
    `SELECT token_id, user_id::text, expires_at, used_at
     FROM password_reset_tokens WHERE token = $1`,
    token
  );

  if (!rows.length) throw new AppError('Invalid or expired reset link', 400, 'INVALID_RESET_TOKEN');

  const row = rows[0];
  if (row.used_at)                        throw new AppError('This reset link has already been used', 400, 'TOKEN_USED');
  if (new Date(row.expires_at) < new Date()) throw new AppError('This reset link has expired. Please request a new one.', 400, 'TOKEN_EXPIRED');

  const passwordHash = await bcrypt.hash(newPassword, BCRYPT_ROUNDS);

  await prisma.user.update({ where: { userId: row.user_id }, data: { passwordHash } });
  await prisma.$executeRawUnsafe(
    `UPDATE password_reset_tokens SET used_at = now() WHERE token_id = $1`,
    row.token_id
  );
}

module.exports = {
  register,
  login,
  verifyMfa,
  sendEmailOtp,
  setupTotp,
  verifyTotpSetup,
  changePassword,
  loginWithSession,
  acceptInvitation,
  getInvitation,
  forgotPassword,
  resetPassword,
};
